ikarus/controlRoutines/adaptiveStepSizing.hh

```diff
@@ -0,0 +1,90 @@
+// SPDX-FileCopyrightText: 2022 The Ikarus Developers mueller@ibb.uni-stuttgart.de
+// SPDX-License-Identifier: LGPL-3.0-or-later
+
+#pragma once
+
+#include <Spectra/SymEigsSolver.h>
+
+#include <Eigen/Core>
+#include <Eigen/Dense>
+#include <Eigen/Sparse>
+
+#include <ikarus/linearAlgebra/nonLinearOperator.hh>
+#include <ikarus/solver/nonLinearSolver/solverInfos.hh>
+#include <ikarus/utils/fancyActivationFunctions.hh>
+
+namespace Ikarus {
+
+  template <typename ActivationFunction = Ikarus::AdaptiveStepSizing::LinearPiecewiseFunction,
+            typename ScalarType         = double>
+  struct DefaultAdaptiveStepSizing {
+    template <typename NonLinearOperator>
+    void operator()(const NonLinearSolverInformation& solverInfo, const NonLinearOperator&,
+                    SubsidiaryArgs& subsidiaryArgs, const ActivationFunction&) {
+      if (targetIterations != 0 && subsidiaryArgs.actStep > 0) {
+        ScalarType actIte       = solverInfo.iterations;
+        subsidiaryArgs.stepSize = sqrt(static_cast<ScalarType>(targetIterations) / actIte) * subsidiaryArgs.stepSize;
+      }
+    }
+    int targetIterations{0};
+  };
+
+  template <typename ActivationFunction = Ikarus::AdaptiveStepSizing::LinearPiecewiseFunction,
+            typename ScalarType         = double>
+  struct FancyAdaptiveStepSizing {
+    template <typename NonLinearOperator>
+    void operator()(const NonLinearSolverInformation& solverInfo, NonLinearOperator& nonLinearOperator,
+                    SubsidiaryArgs& subsidiaryArgs, const ActivationFunction& activationFunction) const {
+      const auto& K = nonLinearOperator.derivative();
+      auto act_step = subsidiaryArgs.actStep;
+
+      auto K_size = K.rows();
+
+      /// Construct matrix operation object using the wrapper class DenseSymMatProd
+      Spectra::DenseSymMatProd<double> op(K);
+
+      /// Construct eigen solver object, requesting the largest three eigenvalues
+      auto num_eigs  = 3;  // TODO: Generalize it
+      auto eigs_conv = 2 * num_eigs;

```

Comment:
This choice is of `eigs_conv`should be elaborated. Refer to https://spectralib.org/doc/classSpectra_1_1SymEigsSolver.html#a469d37e08be076e198c9e383236c5056